### PR TITLE
SQL: fix incorrect ordering of groupings (GROUP BY) based on orderings (ORDER BY)

### LINKED ISSUE #39956: SQL: ordering doesn't work properly for multiple groupings
For a multi `GROUP BY` statement, the order in which the sources are used in the `composite` aggregation is incorrect.

For example, for the query `SELECT first_name AS n, gender AS g FROM test GROUP BY n, g ORDER BY g, n` the generated `composite` aggregation has the order of its sources like the one from the `GROUP BY` definition, where in fact it should be the one from the `ORDER BY` statement:

```
                "sources": [
                    {
                        "1331": {
                            "terms": {
                                "field": "first_name.keyword",
                                "missing_bucket": true,
                                "order": "asc"
                            }
                        }
                    },
                    {
                        "1335": {
                            "terms": {
                                "field": "gender",
                                "missing_bucket": true,
                                "order": "asc"
                            }
                        }
                    }
                ]
```

### PR DESCRIPTION
**Pull request by astefan**
When the `composite` aggregation is being built for the elements under `GROUP BY`, the order in which the `sources` are used should be the same order as the elements of the `ORDER BY` statements. This PR fixes two separate issues:
- when using aliases in the aggregates and GROUP BY/ORDER BY, matching elements from GROUP BY and ORDER BY together doesn't work and the sorting is incorrect
- if functions are used in ORDER BY/GROUP BY, they don't match together and the ordering is not correct.

Fixes https://github.com/elastic/elasticsearch/issues/39956.

### CODE REVIEW

**Comment by astefan** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample

**Comment by astefan** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample
@elasticmachine run elasticsearch-ci/bwc

**Comment by astefan** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample
@elasticmachine run elasticsearch-ci/2

**Review by costin — APPROVED** [review]
LGTM

**Comment by astefan** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample
@elasticmachine run elasticsearch-ci/2

File: x-pack/plugin/sql/qa/src/main/resources/agg-ordering.sql-spec [hunk 1]
```diff
@@ -85,3 +85,36 @@ SELECT emp_no, MIN(languages) AS min FROM test_emp GROUP BY emp_no ORDER BY MAX(
 
 aggWithMixOfOrdinals
 SELECT gender AS g, MAX(salary) AS m FROM test_emp GROUP BY gender ORDER BY 2 DESC LIMIT 3;
+
+multipleGroupingsAndOrderingByGroups_1
+SELECT gender AS g, first_name, last_name AS l FROM test_emp GROUP BY first_name, g, last_name ORDER BY gender, first_name, last_name;
+
+multipleGroupingsAndOrderingByGroups_2
+SELECT gender AS g, first_name, last_name AS l FROM test_emp GROUP BY first_name, g, last_name ORDER BY gender, first_name, last_name;
```
Comment (matriv) [inline x-pack/plugin/sql/qa/src/main/resources/agg-ordering.sql-spec]: This is identical with multipleGroupingsAndOrderingByGroups_1.

**Review by matriv — APPROVED** [review]
LGTM. Left a comment

File: x-pack/plugin/sql/qa/src/main/resources/agg-ordering.sql-spec (hunk 1 above)
Comment (astefan) [reply x-pack/plugin/sql/qa/src/main/resources/agg-ordering.sql-spec]: Good catch.

**Pull request merged**